# 数学规划模型

**优化问题**一般是指用“最好”的方式，使用或分配如劳动力，原材料，设备，资金等有限的资源，使得投入成本最小或者获利最大，其可以归结为如下优化模型：

\begin{equation}
min(max) \quad z=f(x)\\
\text{s.t.}  \quad g_i(x)\leq0,i=1,2,..m
\label{equ1}\tag{1}
\end{equation}

其中，$\mathbf{x}=\{x_1,x_2,...x_n\}$表示决策变量，$f(x)$表示目标函数，$g_i(x)\leq 0(i=1,2,...,m)$表示约束条件。约束条件决定了决策变量$\mathbf{x}$的允许取值范围，即$\mathbf{x}\in\Omega$，其中$\Omega$称为(\ref{equ1})解的可行域

当打算用**数学规划**的方法来处理一个**优化问题**的时候
+ 1 首先要确定**优化目标**是什么，决定优化目标需寻求的决策是什么，决策收到哪些条件的限制
+ 2 再次用数学工具表示出来，即建立数学规划模型
+ 3 最后利用数学规划的方法和相应软件求解，并对结果做一些定性，定量的分析和必要的检验

> 根据目标函数和约束条件的形式，可将数学规划模型分为<br>**1.线性规划模型，2.整数规划模型，3.非线性规划模型，4.目标规划模型和5.动态规划模型**   等

##  线性规划模型

当(1)中的目标函数$f(x)$和约束条件中的$g_i(x)(i=1,2,...m)$均为线性函数时，称模型（1）为线性规划模型，线性规划模型可以具体表示为：

\begin{equation}
min(max) \quad z=\sum_{i=1}^nc_ix_i\\
\text{s.t.}  \quad \sum_{j=1}^na_{ij}x_j\leq b_i,i=1,2,..m
\label{equ2}\tag{2}
\end{equation}

矩阵形式为：

\begin{equation}
min(max) \quad z=\mathbf{c}^T\mathbf{x}\\
\text{s.t.} \quad \mathbf{Ax}\leq \mathbf{b},i=1,2,..m
\label{equ3}\tag{3}
\end{equation}

其中，$\mathbf{x}=\{x_1,x_2,...x_n\}^T$为决策向量；$\mathbf{c}=\{c_1,c_2,...c_n\}^T$为目标函数的系数向量，$\mathbf{b}=\{b_1,b_2,...b_n\}^T$为常数向量，$A=(a_{ij})_{m\times n}$为系数矩阵

### 线性规划模型的建立

**例1.1** 任务分配问题

**问题**： 某车间有甲，乙两台机床，可用于加工3种工件。假定这两台车床的可用台时数分别为800和900，3种工件的数量分别为400,600和500，且已知用2种不同车床加工单位数量不同的工件所需的台时数和加工费用如表1.1所示。

| 车床类型 |  单位加工所需加工台时数   |  单位加工的加工费用       | 可用台时数 |
|----------|------------------------|--------------------|------------|
 | -   |   工件1  &  工件2&  工件3 | 工件1 &工件2 &工件3 |            |
| 甲       | 0.4 & 1.1&  1.0   | 13  &9& 10    | 800        |
| 乙       | 0.5  &  1.2 &  1.3   | 11 &12 &8   | 900        |

**问：** 怎样分配车床的加工任务，才能既满足加工工件的要求，又使加工费用最低？

**问题分析：**
+ 这个优化问题的目标是使加工费用最低
+ 要做的决策是根据甲乙机床加工三种工件的加工时间和费用来分配两种机床对三种工件的加工任务，使得在满足任务的情况下，加工费用最低。
**建模步骤**：

（1）寻求决策

要回答的是，给两种机床的任务分配如何以及加工费用为多少

（2）决定决策变量

本问题中，决定决策（加工费用）的是两种机床分别将加工三种工件的数量，所以设分配给甲机床的加工三种工件的数量分别为$x_1,x_2,x_3$，同理，分配给乙的设为$x_3,x_4,x_5$

（3）确定优化目标

本问题的目标是使加工费用最低，则目标函数可表示为：

$$z=13x_1+9x_2+10x_3+11x_4+12x_5+8x_6$$

（4）寻找约束条件

不难得知两个机床关于三种工件的加工数量的约束是3个**等式约束**，机床的可用台时数是2个**不等式约束**，此外，根据问题实际，各决策变量还有**非负约束**

（5）构成数学模型

将目标函数和约束条件放在一起，就得到了数学模型：
\begin{equation}
min \quad z=13x_1+9x_2+10x_3+11x_4+12x_5+8x_6  \\
s.t. =\left\{
\begin{aligned}
x_1 +x_4& = &400  \\
x_2 +x_5& = &600  \\
x_3 +x_6& = &500  \\
0.4x_1+1.1x_2+x_3 &\leq& 800 \\
0.5x_4+1.2x_5+1.3x_6 &\leq& 900  \\
x_i \geq 0,i=1,2,...6
\end{aligned}
\right.
\end{equation}

### 线性规划模型的求解

python求解线性规划的函数：[scipy.optimize.linprog](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html#scipy.optimize.linprog)

+  minimize a linear objective function subject to linear equality and inequality constraints. 根据受到的线性等式和不等式约束来最小化线性目标函数

+ 其数学求解形式如下：
\begin{equation}
min  \quad c^Tx  \\
s.t. =\left\{
\begin{aligned}
\mathbf{Ax} & \leq & b  \\
A_{eq}x & = & b_{eq}  \\
l<x<u
\end{aligned}
\right.
\end{equation}

其中$x$是decision variables（决策变量），$c,b,b_{eq},l,u$都是向量,$A,A_{eq}$都是矩阵. 

python的线性规划实现和matlab很像,只有一个bound参数需要注意，python中，bound是使用(min,max)这样的元组对的序列，即，每个变量对应一个这样的数对。
使用None表示没有上下阈限制。**默认：上下限是(0,None)即所有的决策变量都是非负的。如果只提供了一个(min，max)对，那么这个min和max会用于所有的决策变量**

In [1]:
import numpy as np
import scipy
from scipy import optimize

# 定义所需的变量
c=np.array([13,9,10,11,12,8])
A=np.array(
  [[0.4,1.1,1,0,0,0],
   [0,0,0,0.5,1.2,1.3]
  ])
b=np.array([800,900])
Aeq=np.array([
    [1,0,0,1,0,0],
    [0,1,0,0,1,0],
    [0,0,1,0,0,1]
])
beq=np.array([400,600,500])

# 调用函数计算
res=scipy.optimize.linprog(c=c,A_ub=A,b_ub=b,A_eq=Aeq,b_eq=beq,bounds=((0,None)))

#查看函数的返回值，res.x返回的就是决策变量的值， res.fun返回的就是当前目标函数的值（决策变量计算后的值对应的目标函数，也就是优化目标的结果）

In [2]:
res.x

array([  0., 600.,   0., 400.,   0., 500.])

In [3]:
res.fun

13800.0

**结果：**

即在甲机床上加工600个工件2，在乙机床上加工400个工件1和500个工件3，可以使得花费最少，为13800

### 模型分析与评价

#### 灵敏性分析

目前没有找到相应的python实现

### 可以转换为线性规划的问题

其实就是一些数学上的变换

**例1.2** 一个工厂的甲乙丙三个车间生产同一种产品，每个产品由4个零件A和3个零件B组成，这两种零件耗用两种不同的原材料，而这两种原材料的现有数额分别是300公斤和400公斤。每个车间的原材料耗用和零件产量如下表所示，问：这三个车间应该各开多少班，才能使这种产品的配套数最大。

| 车间 | 每班用料数  | 每班产量（个） |
|------|-------|-------|
|    -  | 原料1&原料2 | 零件A&零件B    |
| 甲   | 8&6         | 7&5            |
| 乙   | 5&9         | 6&9            |
| 丙   | 3&8         | 8&4            |

**建模步骤**：

和上面一样，分别进行**寻求决策，决定决策变量，确定优化目标，寻找约束条件，构建数学模型**

+ 设甲乙丙三个车间的班数为$x_1,x_2,x_3$,
+ 优化目标是使产品配套数最大，甲乙丙生产A零件的是$7x_1+6x_2+8x_3$,生产B零件是$5x_1+9x_2+4x_3$，由于每个产品需要4个A和3个B，所以产品的最大量是$\frac{7x_1+6x_2+8x_3}{4}$和$\frac{5x_1+9x_2+4x_3}{3}$中较小的一个，即$S=min\{\frac{7x_1+6x_2+8x_3}{4},\frac{5x_1+9x_2+4x_3}{3}\}$
+ 目前得到了数学模型：
\begin{equation}
max \quad S=min\{\frac{7x_1+6x_2+8x_3}{4},\frac{5x_1+9x_2+4x_3}{3}\}  \\
s.t. =\left\{
\begin{aligned}
8x_1+5x_2+3x_3 \leq 300 \\ 
6x_1+9x_2+8x_3 \leq 400  \\
x_1,x_2,x_3 \geq 0
\end{aligned}
\right.
\label{equ4}\tag{4}
\end{equation}

可以很明显看出，(\ref{equ4})中的目标函数不是线性函数，所以需要通过适当的变换变成线性函数。。。比较偏近数学了

\begin{equation}
y=min\{\frac{7x_1+6x_2+8x_3}{4},\frac{5x_1+9x_2+4x_3}{3}\}  \\
\text{上式等价于}
\frac{7x_1+6x_2+8x_3}{4} \geq y,\frac{5x_1+9x_2+4x_3}{3}\geq y \\
\end{equation}

\begin{equation}
max  \quad S=y  \\
s.t. =\left\{
\begin{aligned}
7x_1+6x_2+8x_3-4y \geq 0 \\
5x_1+9x_2+4x_3-3y \geq 0 \\
8x_1+5x_2+3x_3 \leq 300 \\ 
6x_1+9x_2+8x_3 \leq 400  \\
x_1,x_2,x_3,y \geq 0
\end{aligned}
\right.
\end{equation}

由于python求解线性规划的标准形式中是求目标函数的最小值，同时不等式约束全部都是 $\leq$,所以对上式进行变换，对目标函数和前两个不等式约束左右两边$ equation\times(-1)$即可,变为

\begin{equation}
min  \quad S^{'}=-y  \\
s.t. =\left\{
\begin{aligned}
-7x_1-6x_2-8x_3+4y \leq 0 \\
-5x_1-9x_2-4x_3+3y \leq 0 \\
8x_1+5x_2+3x_3 \leq 300 \\ 
6x_1+9x_2+8x_3 \leq 400  \\
x_1,x_2,x_3,y \geq 0
\end{aligned}
\right.
\end{equation}

In [4]:
# 然后再进行求解
c=np.array([0,0,0,-1])
A=np.array([
    [-7,-6,-8,4],
    [-5,-9,-4,3],
    [8,5,3,0],
    [6,9,8,0]
])
b=[0,0,300,500]
res=scipy.optimize.linprog(c=c,A_ub=A,b_ub=b,bounds=((0,None)))

In [5]:
res.x

array([ 15.12319456,  15.71792693,  33.47493628, 116.99235344])

**结果**：

甲乙丙的班数分别为 15.1232,15.7189,33.4749，生产的产品配套数最大，为116.9924套。

**注意**：
一般而言，上述问题的结果应该是整数，所以该线性规划的决策变量应还原为整数，也就是需要增加一个整数约束。

## 整数规划模型

**Integer Programming**，相应的，**线性规划**的英文是**Linear programming**，所以**整数线性规划**的英语是**ilp/ILP**

当（1）中的决策变量$x_i,(i=1,2,...n)$(部分或全部)均为整数时称模型为 **整数规划模型**。

整数规划模型可以是线性的，也可以是非线性的，其大致可分为三类：
+ 当变量全部被限制要求为整数时（即所有决策变量根据问题实际，其数据类型需要为整数），称为完全（纯）整数规划模型
+ 当只需要部分变量被限制为整数时，称为混合整数规划模型
+ 当变量只能取0或者1时，称为0-1规划模型。
+ 当整数规划模型为线性规划模型时，称为整数线性规划模型。

**注意**:
看起来好像求解整数规划模型时，只需要把已得的非整数解去掉非整数部分化为整数就可以了，实际上化整后得到的解不一定是可行解和最优解，除非原线性规划的最优解本来就全是整数，则整数规划的最优解与线性规划的最优解一致，此时可以用求解线性规划的方法求解整数规划。

不过，大多数整数规划模型，需要寻求特殊的解法，由于至今尚未找到一般的求解整数规划模型多项式解法，所以这里只针对一些经典的整数规划问题，讨论建模过程和相应的计算机算法。



根据[Stack Overflow的回答](https://stackoverflow.com/questions/39101137/how-can-i-get-integer-solutions-with-scipy-optimize-linprog),可以知道，无法直接通过scipy.optimize.linprog加上整数限制来得到整数规划模型，需要借用别的python库来实现。。例如：
[Pulp](https://pypi.org/project/PuLP/),[Pyomo](http://www.pyomo.org/),[CVXOPT](http://cvxopt.org/),网上搜索了一下，CVXOPT用的稍微多一点

### CVXOPT库简单认识

cvxpy库 [参考链接1](https://towardsdatascience.com/integer-programming-in-python-1cbdfa240df2),[对应的中文翻译页面](http://www.360doc.com/content/18/0515/07/47919125_754023581.shtml),<br>
cvxopt库 [线性规划相关的solver](http://cvxopt.org/userguide/coneprog.html#optional-solvers),
[GLPK](http://www.gnu.org/software/glpk/glpk.html),
[代码实例](https://nbviewer.jupyter.org/gist/linuxster/7295256),
[CVXOPT matrix 定义](https://cvxopt.org/userguide/matrices.html)

### 应用解决问题

In [38]:
from cvxopt.glpk import ilp
import numpy as np
from cvxopt import matrix

#依旧是上面的问题，但是要加入整数限制。 使用help(ilp)查看这个函数的用法
# 注意， 在调用ilp时， G must be a 'd' matrix  所以最好还是直接使用cvxopt自带的定义数据功能
#ilp算法中除了I B两个参数，其余都要求是matrix类型

#写法1 改对了
# c=matrix([0,0,0,-1],(4,1),tc='d')
# G=matrix([-7,-5,8,6, -6,-9,5,9, -8,-4,3,8, 4,3,0,0],(4,4),tc='d')
# h=matrix([0,0,300,500],(4,1),tc='d')

#这里很奇怪，c和G需要用numpy定义成float类型，常数向量可以直接用matrix定义d类型  tc typecode，
# 不然产生的status是'LP relaxation is dual infeasible'，x（要求的决策向量）就是空
# 原因是  matrix的定义和numpy中不一样,matrix中的定义与numpy刚好是转置的关系，改成上面那样就好了，
# 所以为了方便自己的习惯，还是用下面这个好一些

c=matrix(np.array([0,0,0,-1],dtype=float))
G=matrix(np.array([
    [-7,-6,-8,4],
    [-5,-9,-4,3],
    [8,5,3,0],
    [6,9,8,0]],dtype=float))
h=matrix([0,0,300,500],tc='d')

#4个整数变量，所以I应该是4个整数 
(status,x)=ilp(c=c,G=G,h=h,A=None,b=None,I=set([0,1,2,3]))

In [39]:
c.size,G.size,h.size

((4, 1), (4, 4), (4, 1))

In [7]:
status

'optimal'

In [8]:
print(x)
print(x[0],x[1],x[2],x[3])

[ 1.50e+01]
[ 1.60e+01]
[ 3.30e+01]
[ 1.16e+02]

15.0 16.0 33.0 116.0


```python
(status, x) = ilp(c, G, h, A, b, I, B)
    
    PURPOSE
    Solves the mixed integer linear programming problem
    
        minimize    c'*x
        subject to  G*x <= h
                    A*x = b
                    x[k] is integer for k in I
                    x[k] is binary for k in B
```
所以很明显，c是系数矩阵，x是决策向量，A是等式约束，b是常数向量，G是不等式约束，h是对应的不等式常数向量。

I是整数变量索引的集合，B是二进制变量索引的集合（用来进行0-1 ILP问题）。

### Pulp库

由于下面这个整数规划中没有不等式约束，无法使用cvxopt，所以根据[stackoverflow的回答](https://stackoverflow.com/questions/15029145/using-cvxopt-for-lp-problems-with-only-aeq-beq-no-constraints-with-ax-b),这个人遇到了和我一样的问题，遇到了一个没有不等式约束的01规划问题，所以转而去使用[pulp库](https://github.com/coin-or/pulp)或pyglpk了

+ [pulp IP问题](https://pythonhosted.org/PuLP/pulp.html#module-pulp)
+ [数独问题01规划](https://pythonhosted.org/PuLP/CaseStudies/a_sudoku_problem.html?highlight=example)
+ [运输分配问题 线性规划](https://pythonhosted.org/PuLP/CaseStudies/a_transportation_problem.html?highlight=example)
+ [pulp案例学习](https://pythonhosted.org/PuLP/CaseStudies/)

### 0-1整数规划（没有不等式约束）

在一些决策问题中，当每个需要做的决策只有两种时，可以使用0-1整数规划来建模。而0-1整数规划是整数规划中的特殊情形，它的决策变量
$x_i$只取值0或者1

**例1.3** 指派问题

**问题：** 拟分配$n$人去干$n$项工作，每人干且仅干一项工作，若分配第$i$人去干第$j$项工作，需花费$c_{ij}$单位时间，问应如何分配工作才能使工人花费的总时间最少

**问题分析：**在该问题中，每一个人均能从事所有的工作，只是从事不同的工作需要花费的单位时间不同而已。由于每个人干且仅干一项工作，所以第$i$人和第$j$项工作之间，只存在干与不干的关系，因此，可以用1和0来刻画第$i$人和第$j$项工作之间干与不干的关系，进而建立0-1规划模型

**建模步骤：**

（1）寻求决策：工人花费的总时间最少

（2）决策变量：由于每个人和每项工作之间均需建立关系，所以用变量$x_{ij}$来描述第$i$人和第$j$项工作之间干与不干的关系，
$x_{ij}$=1，干，$x_{ij}=0$ 不干

（3）确定优化目标：

对于第$i$人，若被分配去干第$j$个工作，此时$x_{ij}=1$,所需时间为$c_{ij}$,所不被分配去干第$j$项工作，则需要时间为0，所以第$i$人分配干第$j$项工作的时间可以用$c_{ij}x_{ij}$来表示。

由于在所有工作中，第$i$个人干且仅干一项工作，假设第$i$个人去干第$j_0$项工作，当$j \neq j_0$,则花费的时间为 $c_{ij}x_{ij}=0$,所以第$i$人总共花费的单位时间为 $c_{ij_0}x_{ij_0}=\sum_{j=1}^nc_{ij}x_{ij}$,所以所有人花费的总时间是：
$T=\sum_{i=1}^n\sum_{j=1}^nc_{ij}x_{ij}$, $n$个工作，$n$个人

（4）约束条件

因为是$n$个工作$n$个人且每人干且仅干一个工作,所以每个人要干一个工作，即对第$i$个人而言，$\sum_{j=1}^nx_{ij}=1$,（这个人只干一个工作），对第$j$项工作而言，$\sum_{i=1}^nx_{ij}=1$（这个工作只有一个人干）

（5）构成数学模型
\begin{equation}
min \quad T=\sum_{i=1}^n\sum_{j=1}^nc_{ij}x_{ij} \\
s.t.=\left\{
\begin{aligned}
\sum_{j=1}^nx_{ij}=1 (i=1,2,...,n)\\
\sum_{i=1}^nx_{ij}=1 (j=1,2,...,n) \\
x_{ij}=0 \quad or \quad 1(i,j=1,2,...,n) \\
\end{aligned}
\right.
\end{equation}

不难看出，解决一个指派问题（0-1整数规划问题），只需要给出一个矩阵$\mathbf{C}=c_{ij}$,$\mathbf{C}$被称为指派问题的系数矩阵

这里假设给出的矩阵为：
\begin{equation}
\mathbf{C}=
\begin{pmatrix} 
3&8&2&10&3\\
8&7&2&9&7\\
6&4&2&7&5\\
8&4&2&3&5\\
9&10&6&9&10
\end{pmatrix}
\end{equation}

也就是5个人，5个工作,其实要得到25个决策变量。。。由于$x_{ij}(i,j=1,2,3,4,5)$是二维决策变量，要求解的话要变成一维决策变量，$y_k(k=1,2,...25)$

**等式约束**

$A_{eq}x=beq$ 这里$x$决策变量是25d，$beq$是10，所以不难推断出$A_{eq}$是$10\times25$d,另外，上面的两个等式约束其实可以用一个式子来表示，$x_{ij}$是一个25d的变量，其实把式子全部列出来就明白了

当i=0-4时,5个式子
\begin{equation}
\text{i=0} \quad 1x_{11}+1x_{12}+1x_{13}+1x_{14}+1x_{15}+0x_{21}+...+0x_{25}+0x_{31}+...+0x_{35}...+0x_{55}=1  \\
\text{i=1} \quad 0x_{11}...+0x_{15}+1x_{21}+1x_{22}+1x_{23}+1x_{24}+1x_{25}+0x_{31}+...+0x_{35}...+0x_{55}=1   
\end{equation}

找规律，系数为1的决策变量的序号
+ 行序号直接就是i,
+ 列序号范围分别是：0-4，5-9,10-14....则可以表示为$[i,5i:5i+4]$

同理，当$j=0-4$时，也是五个式子

\begin{equation}
1x_{11}+0x_{12}+...+0x_{15}+1x_{21}+0x_{22}..+0x_{25}+...+1x_{51}+0x_{52}+0x_{55}=1
\end{equation}

找规律，则系数为1的决策变量的序号
+ 行序号需要加5，即$i+5$
+ 列序号，
    + j=0时，numpy中对应列序号是5，列的系数值是 1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 0
    + j=1时，numpy中对应列序号是6，列的系数值是 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0
    + 所以后五行的取值规律是： [i+5,i:25:5]

在 [在线的matlab网站](https://octave-online.net/)中输入：

```
for i=1:5
    a(i,(i-1)*5+1:5*i)=1;
    a(5+i,i:5:25)=1;
end
```
就可以看到a的全貌了,这里注意matlab和numpy有个很大的区别就是，
+ matlab的切片是[start:step:end]  且开始和结束都是闭
+ numpy的切片，开始:结束:步长   [start,end)是左闭右开

可以对比A[0:5,:]和在线matlab中输入 a(1:5,:),以及A[5:10,:]和 a(6:10,:)

In [59]:
# 无法得到结果的代码。。。错误，放弃
c=matrix(np.array([[3,8,2,10,3],[8,7,2,9,7],[6,4,2,7,5],[8,4,2,3,5],[9,10,6,9,10]]).reshape(25,1),tc='d')
A=np.zeros(shape=(10,25),dtype=float)

for i in range(0,5):
    A[i,i*5:5*i+5]=1  #0 1 2 3 4 行
    A[5+i,i:25:5]=1  # 5 6 7 8 9 行
    
A=matrix(A)
b=matrix(np.ones(shape=(10,1)),tc='d')

#很明显这个问题中，是没有不等式约束的，所以G=None，h=None，但是如果不写G和h就会报错，所以填一个空的矩阵上去就可以
# G*x <= h 根据这个，很明显 G全为1的时候，由于x最大就是1，所以肯定<=1 
G=matrix(0,(10,25),tc='d')
h=matrix(0,(10,1),tc='d')

(status,x)=ilp(c=c,A=A,b=b,B=set(range(10)))
# 然鹅  status 'LP relaxation is dual infeasible'

In [66]:
import pulp

### 另一个0-1整数规划的例子(有不等式约束和等式约束)

**例1.4** 某公司有5个项目被列入投资计划，各项目的投资额和期望的投资收益如下表所示，该公司只有600百万资可用于投资，由于技术上的原因投资受到以下约束：

（1）在项目1,2和3中必须有一项被选中

（2）项目3和项目4只能选中一项

（3）项目5被选中的前提是项目1必须被选中    即$x_5\leq x_1$可以化为$-x_1+x_5\leq 0$

如何在上述条件下选择一个最好的投资方案，使投资收益最大？

| 项目 | 投资额（百万元） | 投资收益（百万元） |
|------|------------------|--------------------|
| 1    | 210              | 150                |
| 2    | 300              | 210                |
| 3    | 100              | 60                 |
| 4    | 130              | 80                 |
| 5    | 260              | 180                |

**问题分析：** 该问题是个投资决策问题，由于每个项目的投资额和投资收益是固定的，因此对每个项目而言，不存在投资某个项目多少个的问题，只存在是否被选中的问题，选/不选 这种两元问题就可以使用0-1规划问题

**建模步骤**

（1）寻求决策： 使投资收益最大

（2）确定决策变量：每个项目选或者不选，设五个项目选或不选分别为$x_i$， $x_i=1$表示选，$x_i=0$表示不选

（3）确定优化目标: $max \quad S=150x_1+210x_2+60x_3+80x_4+180x_5$

（4）约束条件：投资额一共600百万，（1）（2）（3）个表示一个约束

（5）构成数学模型：

\begin{equation}
max\quad S=150x_1+210x_2+60x_3+80x_4+180x_5 \\
\text{s.t.}=\left\{
\begin{aligned}
210x_1+300x_2+100x_3+130x_4+260x_5\leq 600\\
x_1+x_2+x_3=1\\
x_3+x_4\leq1\\
-x_1+x_5\leq0
\end{aligned}
\right.
\end{equation}

In [44]:
c=matrix(np.array([-150,-210,-60,-80,-180],dtype=float))
# 不能np.array(-1*[150,210,60,80,180],dtype=float)，否则c.size=(0,1)就不对了

G=matrix(np.array([[210,300,100,130,260],[0,0,1,1,0],[-1,0,0,0,1]]),tc='d')
h=matrix([600,1,0],tc='d')
A=matrix(np.array([1,1,1,0,0]).reshape(1,5),tc='d')
b=matrix([1],tc='d')

(status,x)=ilp(c=c,G=G,h=h,A=A,b=b,B=set(range(5)))

维度方面 $c'$是转置，也就是说x应该基本都是$n*1$

In [45]:
A.size, b.size,G.size,h.size,c.size

((1, 5), (1, 1), (3, 5), (3, 1), (5, 1))

In [46]:
status

'optimal'

In [48]:
print(x)

[ 1.00e+00]
[ 0.00e+00]
[ 0.00e+00]
[ 1.00e+00]
[ 1.00e+00]



**结果分析：** 

投资第一第四和第五个项目时可以得到最大收益（ilp不会直接给出目标函数的最优值，而只会给出一组最优解。。需要自己算）
150+80+180=410